execute.go

```diff
@@ -94,7 +94,10 @@ func (executor *ParallelExecutor) Execute(ctx *ExecutionContext) (map[string]int
 		for {
 			select {
 			// we have a new result
-			case payload := <-resultCh:
+			case payload, ok := <-resultCh:
+				if !ok {
+					return
+				}
```

Comment:
It's not clear to me how to introduce a test for this scenario. It does seem like restructuring channels/wait groups/locks may be in order for a more modular and testable execution flow.